lib/spack/spack/compilers/__init__.py

```diff
@@ -134,90 +204,109 @@ def _init_compiler_config(*, scope):
+        init_config: if True, initialize configuration files if no compilers are available
     """
-    compiler_config = get_compiler_config(scope, init_config)
+    compiler_cfg = compiler_config(scope=scope, init_config=init_config)
     for compiler in compilers:
-        compiler_config.append(_to_dict(compiler))
-    global _cache_config_file
-    _cache_config_file = compiler_config
-    spack.config.set("compilers", compiler_config, scope=scope)
+        compiler_cfg.append(_to_dict(compiler))
+    spack.config.set("compilers", compiler_cfg, scope=scope)
 
 
-@_auto_compiler_spec
-def remove_compiler_from_config(compiler_spec, scope=None):
-    """Remove compilers from the config, by spec.
+def remove_compiler_from_config(
+    compiler_spec: Union[str, spack.spec.CompilerSpec], *, scope: Optional[str] = None
+) -> bool:
+    """Remove compilers from configuration by spec.
 
-    Arguments:
-        compiler_specs: a list of CompilerSpec objects.
-        scope: configuration scope to modify.
+    If scope is None, all the scopes are searched for removal.
+
+    Args:
+        compiler_spec: compiler to be removed
+        scope: configuration scope to modify
+
+    Returns:
+        True if any compiler has been removed, False otherwise
     """
-    # Need a better way for this
-    global _cache_config_file
+    if not isinstance(compiler_spec, spack.spec.CompilerSpec):
+        compiler_spec = spack.spec.CompilerSpec(compiler_spec)
 
-    compiler_config = get_compiler_config(scope)
-    config_length = len(compiler_config)
+    candidate_scopes = [scope]
+    if scope is None:
+        candidate_scopes = list(spack.config.config.scopes.keys())
 
+    removal_happened = False
+    for current_scope in candidate_scopes:
+        removal_happened |= _remove_compiler_from_scope(compiler_spec, scope=current_scope)
+
+    return removal_happened
+
+
+def _remove_compiler_from_scope(compiler_spec, scope):
+    """Removes a compiler from a specific configuration scope.
+
+    Args:
+        compiler_spec: compiler to be removed
+        scope: configuration scope under consideration
+
+    Returns:
+         True if one or more compiler entries were actually removed, False otherwise
+    """
+    assert scope is not None, "a specific scope is needed when calling this function"
+    compiler_cfg = compiler_config(scope=scope)
     filtered_compiler_config = [
-        comp
-        for comp in compiler_config
+        compiler_entry
+        for compiler_entry in compiler_cfg
         if not spack.spec.parse_with_version_concrete(
-            comp["compiler"]["spec"], compiler=True
+            compiler_entry["compiler"]["spec"], compiler=True
         ).satisfies(compiler_spec)
     ]
 
-    # Update the cache for changes
-    _cache_config_file = filtered_compiler_config
-    if len(filtered_compiler_config) == config_length:  # No items removed
-        CompilerSpecInsufficientlySpecificError(compiler_spec)
```

Comment:
This is a bug. Here we are not raising the exception object, and on top of that the exception is for multiple matches, while in this code path we have no match